README.md

timer-stopwatch [![Codeship Status for MickCrozier/timer-stopwatch](https://codeship.io/projects/d68910a0-e989-0131-eba0-368dc75eab9e/status)](https://codeship.io/projects/26196)
========================

A stopwatch and countdown clock module for node.js


Install
======

```shell
npm install timer-stopwatch
```


How to
======
**The Gist**
new Stopwatch([countDownMS], [options])


Basic timers
```js
var Stopwatch = require('timer-stopwatch');

var timer = new Stopwatch(60000); // A new countdown timer with 60 seconds
var stopwatch = new Stopwatch(); // A new count up stopwatch. Starts at 0.
```

###Events
```js
// Fires every 50ms by default. Change setting the 'refreshRateMS' options
timer.on('time', function(time) {
	console.log(time.ms); // number of milliseconds past (or remaining);
});

// Fires when the timer is done
timer.on('done', function(){
	console.log('Timer is complete');
});

// Fires when the timer is almost complete - default is 10 seconds remaining. Change with 'almostDoneMS' option
timer.on('almostdone', function() {
	console.log('Timer is almost complete');
});

```

###Methods
```js
timer.start();
timer.stop();
timer.startstop(); // toggle the state of the timer. (If started, stop it. If stopped, start it.)
timer.reset(countDownMS); // optional countDownMS to reset countdown to that many milliseconds
```

###Properties
```js
timer.ms;		// Number of milliseconds on the clock
```



###Options
```js
var options = {
	refreshRateMS: 10,		// How often the clock should be updated
	almostDoneMS: 10000, 	// When counting down - this event will fire with this many milliseconds remaining on the clock
}

var timer = new Stopwatch(60000, options);
```



Testing
======

Unit and Integration tests (requires dev dependencies)
```shell
npm test
```

License
======
MIT License



CHANGELOG.md
Changelog

======
####0.1.1
- Removed clock property and time formatting. Convienience was good, but got confusing at application level. Belongs in the view.

- Reset now optionally accepts a millisecond value to reset the clock to.

####0.1.4
- Bug Fixes

####0.1.5
- Emit time event on reset

####0.1.6
- Emit forcestop when the clock is stopped

test/unit/Stopwatch.spec.js

/////////////////////////////////////////////////////
var expect = require('expect.js');			/////////
/////////////////////////////////////////////////////




describe('Countdown Timer', function() {
	
	var Stopwatch = null;

	beforeEach(function() {
		Stopwatch = require('../../lib/Stopwatch.js');
	});
	
	afterEach(function() {
		
	});

	it('should be able to create a countdown watch with 30 seconds', function() {
		var countdownTimer = new Stopwatch(30000);
		expect(countdownTimer.countDownMS).to.be(30000);
		expect(countdownTimer.ms).to.be(30000);
	});

	it('should be able to change the countdown time during reset', function() {
		var countdownTimer = new Stopwatch(30000);
		expect(countdownTimer.countDownMS).to.be(30000);
		countdownTimer.reset(60000);
		expect(countdownTimer.countDownMS).to.be(60000);
		expect(countdownTimer.ms).to.be(60000);
	});


	it('should countdown with a normal refresh rate', function(done) {
		var countdownTimer = new Stopwatch(60000);
		var startTime = countdownTimer.ms;
		countdownTimer.start();
		setTimeout(function(){
			expect(countdownTimer.ms).to.be.above(59998);
		}, 10);

		setTimeout(function(){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.be.below(startTime);
			done();
		}, 70);
	});

	it('should countdown at max refresh speed', function(done) {
		var countdownTimer = new Stopwatch(60000, {refreshRateMS:1});
		var startTime = countdownTimer.ms;
		countdownTimer.start();
		setTimeout(function(){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.be.below(startTime);
			done();
		}, 3);
	});


	it('should countdown, pause, then continue', function(done) {
		var countdownTimer = new Stopwatch(50, {refreshRateMS:1});
		var startTime = countdownTimer.ms;
		var splittime = 0;
		countdownTimer.start();
		setTimeout(function(){
			countdownTimer.stop();
			splittime = countdownTimer.ms;
			expect(splittime).to.be.below(startTime);
		}, 10);

		setTimeout(function(){
			expect(splittime).to.be(countdownTimer.ms);
			countdownTimer.start();
		}, 20);

		setTimeout(function(){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.be.below(splittime);
			done();
		}, 30);
	});


	it('should fire the time event', function(done) {
		var countdownTimer = new Stopwatch(60000);
		var startTime = countdownTimer.ms;
		countdownTimer.start();
		countdownTimer.on('time',function(time){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.be.below(startTime);
			expect(time.ms).to.equal(countdownTimer.ms);
			done();
		});
	});

	it('should fire the almostdone event', function(done) {
		var countdownTimer = new Stopwatch(40, {almostDoneMS:20, refreshRateMS:10});
		var startTime = countdownTimer.ms;

		var onDone = function onDone(){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.below(20);
			expect(countdownTimer.ms).to.above(5);
			countdownTimer.removeListener('almostdone', onDone);
			done();
		};

		countdownTimer.start();
		countdownTimer.on('almostdone', onDone);
	});

	it('should fire the done event', function(done) {
		var countdownTimer = new Stopwatch(30, {almostDoneMS:20});
		var startTime = countdownTimer.ms;

		var onDone = function onDone(){
			countdownTimer.stop();
			expect(countdownTimer.ms).to.equal(0);
			countdownTimer.removeListener('done', onDone);
			done();
		};

		countdownTimer.start();
		countdownTimer.on('done', onDone);
	});

	it('should fire the done event when done again after reset', function(done) {
		var countdownTimer = new Stopwatch(40, {almostDoneMS:20, refreshRateMS:10});
		var startTime = countdownTimer.ms;

		var doneFiredTimes = 0;

		setTimeout(function() {
			expect(doneFiredTimes).to.be(3);
			countdownTimer.removeListener('done', onDone);
			done();
		}, 350);

		var onDone = function onDone(){
			doneFiredTimes++;
			setTimeout(function(){
				expect(countdownTimer.ms).to.equal(0);
				expect(countdownTimer.doneFired).to.equal(true);
				countdownTimer.reset();
				expect(countdownTimer.ms).to.equal(40);
				countdownTimer.start();
				setTimeout(function(){
					expect(countdownTimer.doneFired).to.equal(false);
					expect(countdownTimer.ms).to.be.above(15);
					expect(countdownTimer.ms).to.be.below(31);
				}, 20);
			}, 100);	
		};

		countdownTimer.start();
		countdownTimer.on('done', onDone);
		
	});

	it('should fire the almostdDone event when ALMOST done again after reset', function(done) {
		var countdownTimer = new Stopwatch(40, {almostDoneMS:20, refreshRateMS:10});
		var startTime = countdownTimer.ms;

		var doneFiredTimes = 0;

		setTimeout(function() {
			expect(doneFiredTimes).to.be(3);
			countdownTimer.removeListener('almostdone', onDone);
			done();
		}, 350);


		var onDone = function onDone(){
			doneFiredTimes++;
			setTimeout(function(){
				countdownTimer.reset();
				countdownTimer.start();
			}, 100);
		};

		countdownTimer.start();

		countdownTimer.on('almostdone', onDone);
	});

	it('should fire the forcestop event', function(done) {
		var countdownTimer = new Stopwatch(50, {almostDoneMS:20});

		var onForceStop = function onForceStop(){
			expect(true).to.equal(true);
			countdownTimer.removeListener('forcestop', onForceStop);
			done();
		};

		countdownTimer.start();
		countdownTimer.on('forcestop', onForceStop);
		countdownTimer.stop();
	});
});

describe('Stopwatch', function() {
	
	var Stopwatch = null;

	beforeEach(function() {
		Stopwatch = require('../../lib/Stopwatch.js');
	});
	
	afterEach(function() {
		
	});

	it('should be able to create a stopwatch', function() {
		var countdownTimer = new Stopwatch();
		expect(countdownTimer.countDownMS).to.be(false);
	});


	it('should count up with a moderate refresh rate', function(done) {
		var stopwatch = new Stopwatch();
		var startTime = stopwatch.ms;
		stopwatch.start();
		setTimeout(function(){
			expect(stopwatch.ms).to.be.equal(startTime);
		}, 30);

		setTimeout(function(){
			stopwatch.stop();
			expect(stopwatch.ms).to.be.above(startTime);
			done();
		}, 60);
	});

	it('should countup at max refresh speed', function(done) {
		var stopwatch = new Stopwatch(false, {refreshRateMS:1});
		var startTime = stopwatch.ms;
		stopwatch.start();
		setTimeout(function(){
			stopwatch.stop();
			expect(stopwatch.ms).to.be.above(startTime);
			done();
		}, 3);
	});


	it('should countup, pause, then continue', function(done) {
		var stopwatch = new Stopwatch(false, {refreshRateMS:1});
		var startTime = stopwatch.ms;
		var splittime = 0;
		stopwatch.start();
		setTimeout(function(){
			stopwatch.startstop();
			splittime = stopwatch.ms;
			expect(splittime).to.be.above(startTime);
		}, 10);

		setTimeout(function(){
			expect(splittime).to.be(stopwatch.ms);
			stopwatch.startstop();
		}, 20);

		setTimeout(function(){
			stopwatch.startstop();
			expect(stopwatch.ms).to.be.above(splittime);
			done();
		}, 30);
	});

	it('Should reset to 0', function(done) {
		var stopwatch = new Stopwatch(false, {refreshRateMS:1});

		stopwatch.startstop();

		setTimeout(function(){
			expect(stopwatch.ms).to.be.above(90);
			expect(stopwatch.ms).to.be.below(110);
			stopwatch.reset();
			expect(stopwatch.ms).to.be(0);
			setTimeout(function(){
				stopwatch.startstop();
				setTimeout(function(){
					expect(stopwatch.ms).to.be.above(90);
					expect(stopwatch.ms).to.be.below(110);
					stopwatch.reset();	
					expect(stopwatch.ms).to.be(0);
					done();
				}, 100);
			}, 100);
		}, 100);
	});


	it('should fire the time event', function(done) {
		var stopwatch = new Stopwatch();
		var startTime = stopwatch.ms;
		stopwatch.start();
		stopwatch.on('time',function(time){
			stopwatch.stop();
			expect(stopwatch.ms).to.be.above(startTime);
			expect(time.ms).to.equal(stopwatch.ms);
			done();
		});
	});

	it('should NOT fire the almostdone event', function(done) {
		var stopwatch = new Stopwatch(false, {almostDoneMS:50, refreshRateMS:10});
		stopwatch.start();
		var fired = false;
		stopwatch.on('almostdone',function(formatted, ms){
			fired = true;
		});
		setTimeout(function() {
			expect(fired).to.be(false);
			done();
		}, 60);
	});


	it('should NOT fire the done event', function(done) {
		var stopwatch = new Stopwatch();
		stopwatch.start();
		var fired = false;
		stopwatch.on('done',function(formatted, ms){
			fired = true;
		});
		setTimeout(function() {
			expect(fired).to.be(false);
			done();
		}, 60);
	});

	it('should fire the forcestop event', function(done) {
		var stopwatch = new Stopwatch();

		var onForceStop = function onForceStop(){
			expect(true).to.equal(true);
			stopwatch.removeListener('forcestop', onForceStop);
			done();
		};

		stopwatch.start();
		stopwatch.on('forcestop', onForceStop);
		stopwatch.stop();
	});

});



test/integration/Stopwatch.spec.js
var expect = require('expect.js');

describe('Stopwatch operations', function() {
	var Stopwatch = require('../../index.js');
	beforeEach(function(){

	});

	it('should load up the stopwatch', function() {
		var countdownTimer = new Stopwatch(60);
		expect(countdownTimer).to.be.ok();
	});
});

Gruntfile.js
module.exports = function(grunt) {
	grunt.initConfig({
		pkg: grunt.file.readJSON('package.json'),

		jshint: {
			options: {
				curly: true,
				eqeqeq: true,
				eqnull: true,
				node: true,
				globals: {
					
				},
		    },
		    all: ['*.js', 'lib/**/*.js', 'api/**/*js', 'test/**/*.js'],
		    testu: ['*.js', 'lib/**/*js', 'api/**/*js', 'test/unit/*js'],
		},


		mochaTest: {
	      	unit: {
	        	options: {
	          		reporter: 'spec'
	        	},
	        	src: ['test/unit/*spec.js']
	      	},

	    },


		watch: {
			options: {
				atBegin: true,
			},

			testall: {
				files: ['*.js', 'lib/**/*js', 'api/**/*js', 'test/unit/*js'],
				//tasks: ['jshint:all', 'mochaTest:unit', 'startMongo', 'wait:giveMongoSomeTimeToLoad', 'force:on','mochaTest:integration', 'force:restore', 'stopMongo'],
				tasks: ['jshint:all', 'mochaTest:unit'],
				options: {
					// spawn: false,
				},
			},


			testu: {
				files: ['*.js', 'lib/**/*js', 'api/**/*js', 'test/unit/*js'],
				tasks: ['jshint:testu', 'mochaTest:unit'],
				options: {
					//spawn: false,
				}
			},

		},


	});
	

	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-mocha-test');
	grunt.loadNpmTasks('grunt-contrib-watch');


	grunt.registerTask('test', 'Runs Mocha tests add --watch for continuous', function () {

	    if(grunt.option('watch')) {
      		grunt.task.run([
    			'watch:testall'
    		]);
	    } else {
		    grunt.task.run([
		        'jshint:all',
		        'mochaTest'
		    ]);
	    }
	});

};

lib/Stopwatch.js
// Developed for TKD-Score-Server by Mick Crozier 2014
// MIT License

// Based on work from Kore Byberg - http://www.timpelen.com/extra/sidebars/stopwatch/stopwatch.htm


var EventEmitter = require('events').EventEmitter;
var _ = require('lodash');



function Stopwatch(countDownMS, options) {

	this.runTimer = false;	// Whether the clock should run or not
	this.hasBeenStopped = false;	// signifies the clock should stop	
	this.stoptime = 0;  	// the time the clock has been paused at

	this.refreshTimer = 0;	// timeout interval for counter
	this.done = false;		// true if comleted
	this.almostDoneFired = false;	// true if almostDone event has been fired (prevent mlti hits)
	this.doneFired = false;		// true if done event has been fired (prevent multi hits)

	this.countDownMS = (countDownMS) || false;
	this.ms = this.countDownMS || 0;


	//// options
	if(!options) {options = {};}
	this.refreshRateMS = options.refreshRateMS || 50;
	this.almostDoneMS = options.almostDoneMS || 10000;


	//// init
	this.reset(countDownMS);


}


Stopwatch.prototype = {

	/**
	 * Start the timer
	 */
	start: function() {
		this.startstop('start');
	},

	/**
	 * Stops the timer
	 *
	 * Emits the event forcestop,
	 * with one parameter passed to the callback,
	 * that consists of the elapsed time.
	 */
	stop: function() {
		this.startstop('stop');
	},

	/**
	 * Stop a timer, and reset it to it's defaults.
	 * Change the countdown value, if a paramter is provided.
	 *
	 * @param {Integer} Milliseconds to set the timer to.
	 */

	reset: function(countDownMS) {

		this.stop();
		this.hasBeenStopped = false;
		this.doneFired = false;
		this.almostDoneFired = false;
		this.stoptime = 0;
		this.done = false;
		clearTimeout(this.refreshTimer);


		if(countDownMS) {
			this.countDownMS = countDownMS;
		}
		this.ms = this.countDownMS || 0;

		this.emit('time',{ms: this.ms});
	},

	/**
	 * Toggle the state of the timer.
	 * If one of start or stop is given as a argument to the
	 * function then the timer will be forced into that state.
	 *
	 * If no argument is given, then the timer's state will be toggled
	 * between start and stop.
	 * i.e. The timer will be stopped, if it is running, and the timer
	 * will be started if the timer is already stopped. 
	 *
	 * @param {String} start|stop Optional paramter.
	 * @returns {Boolean} true if the timer is running, false otherwise.
	 */

	startstop: function(force) {
		if(force) {
			if(force === 'start') {
				this.runTimer = false;
			} else if(force === 'stop') {
				this.runTimer = true;
			}
		}

		var startdate = new Date();
		var starttime = startdate.getTime();

		if(!this.runTimer) {

			this.runTimer = true;
			this.counter(starttime);
		} else {
			this.runTimer = false;
			this.hasBeenStopped = true;
			this.emit('forcestop', {elapsed: this.ms});
		}

		return this.runTimer;
	},


	/**
	 * The main function, consisting of the code that interfaces
	 * with the system time interface to find the elapsed time.
	 *
	 * Not intended to be used by end-users.
	 *
	 */
	counter: function(starttime) {

		var currenttime = new Date();
		var timediff = currenttime.getTime() - starttime;

		if(this.hasBeenStopped) {
			timediff = timediff + this.stoptime;
		}


		if(this.runTimer) {

			var time;
			if(this.countDownMS) { // run clock backwards declaring done when reached 0 
				time = this.countDownMS - timediff; 

				if(time < 0) {time = 0;}
				this.ms = time;


				if(time === 0) {
					this.stop(); // stop the clock

					if(!this.doneFired) {
						this.doneFired = true;
						this.done = true;
						this.emit('done');
					}

				} else if (time < this.almostDoneMS) {
					if(!this.almostDoneFired) {
						this.almostDoneFired = true;
						this.emit('almostdone');
					}
				}


			} else { // run clock forwards with no auto stop

				time = timediff; 
				this.ms = time;
			}


			this.emit('time',{ms: this.ms});
			var that = this;

			this.refreshTimer = setTimeout(function(){that.counter(starttime);},this.refreshRateMS);

		} else {
			clearTimeout(this.refreshTimer);
			this.stoptime = timediff;
		}

	},


};


_.extend(Stopwatch.prototype, EventEmitter.prototype);


module.exports = Stopwatch;

